### PR TITLE
[DPE-4487] Add Integration Tests for Azure Storage

### PR DESCRIPTION
**Pull request by theoctober19th**

### CODE REVIEW

File: tests/integration/integration-tests-kyuubi.sh [hunk 1]
```diff
@@ -85,6 +86,61 @@ setup_kyuubi_pod() {
...
+
+  # Create Azure storage container
+  create_azure_container $AZURE_CONTAINER
+
+  storage_account_name=$(get_storage_account)
+  storage_account_key=$(get_azure_secret_key)
+  warehouse_path=$(construct_resource_uri $AZURE_CONTAINER warehouse abfss)
+  file_upload_path=$(construct_resource_uri $AZURE_CONTAINER "" abfss)
+
+  # Write Spark configs inside the Kyuubi container
```
Comment (deusebio) [inline tests/integration/integration-tests-kyuubi.sh]: **suggestion** maybe one could also use the `spark-client.service-account-registry` to setup the configuration, as much as we do elsewhere. It feels to me that one single command with the configs would be more readable and maintainable than a bunch of `kubectl exec` commands

File: tests/integration/integration-tests-kyuubi.sh [hunk 2]
```diff
@@ -85,6 +86,61 @@ setup_kyuubi_pod() {
...
+          /bin/bash -c 'echo spark.hadoop.fs.azure.account.key.$ACCOUNT_NAME.dfs.core.windows.net=$SECRET_KEY >> /etc/spark8t/conf/spark-defaults.conf'
+  kubectl -n $NAMESPACE exec kyuubi-test -- \
+      env WAREHOUSE="$warehouse_path" \
+          /bin/bash -c 'echo spark.sql.warehouse.dir=$WAREHOUSE >> /etc/spark8t/conf/spark-defaults.conf'
+  kubectl -n $NAMESPACE exec kyuubi-test -- \
+      env UPLOAD_PATH="$file_upload_path" \
+          /bin/bash -c 'echo spark.kubernetes.file.upload.path=$UPLOAD_PATH >> /etc/spark8t/conf/spark-defaults.conf'
+
+  # Wait some time for the server to be up and running
+  sleep 10
```
Comment (deusebio) [inline tests/integration/integration-tests-kyuubi.sh]: **suggestion** maybe we could have the classic

```
while (condition):
    sleep 10
```

That we use elsewhere, e.g. in the `wait_for_pod`

**Review by deusebio — COMMENTED** [review]
The code does what it should, but I have a comment to maybe make the tests more maintainable and reusable, possibly also improving in consistency. 

Right now we handle s3 setup and azure setup a bit differently, so for instance: s3 configuration are fed into the spark-submit command, while azure are fed using `kubectl` commands. Also a number of business logic in the iceberg test is just copied and pasted. I'm wondering whether we could rewrite the tests such that it reads:

```
(setup_user_context && setup_object_storage_s3 && test_iceberg_example_in_pod && cleanup_user_success) || cleanup_user_failure_in_pod

(setup_user_context && setup_object_storage_azure && test_iceberg_example_in_pod && cleanup_user_success) || cleanup_user_failure_in_pod
````

The custom part (between the two) is just the `setup_object_storage_*`part, where we both setup using the CLI but also we inject the right configuration using the `spark-client.service-account-registry` that is embedded in the OCI image. 

Then, the iceberg test should be just using these configuration already injected and add the spark-submit command with the iceberg configuration only

It is not too critical, but I would honestly spend some time right now to do this, such that when adding other backends (blob storage, abfs, etc) it should be easier and more straight forward. Also this should be super easy to translate this into more structured tests, where the azure or s3 configuration is setup by the integration-hub

File: tests/integration/integration-tests-kyuubi.sh (hunk 1 above)
Comment (theoctober19th) [reply tests/integration/integration-tests-kyuubi.sh]: I've reduced the bunch of `kubectl exec` commands to just a couple of them. I couldn't come up with a way to do it using lesser `exec` commands.

File: tests/integration/integration-tests-kyuubi.sh (hunk 2 above)
Comment (theoctober19th) [reply tests/integration/integration-tests-kyuubi.sh]: I think we can remove this wait here, this came here when copying the reusable code from Spark tests.

**Review by welpaolo — APPROVED** [review]
LGTM! 

I have only a comment regarding the massive usage of bash. As future work, strongly believe we should starting using the `spark-test` library and move all our tests with pytest

**Review by deusebio — APPROVED** [review]
Great! The tests look really great! thanks!

**edit** I noticed that sql tests is with S3 only. Would it be possible to make it both for S3 and azure?

File: tests/integration/integration-tests.sh [hunk 3]
```diff
@@ -314,11 +262,61 @@ test_iceberg_example_in_pod(){
...
+  if [ $return_value -eq 1 ]; then
+    exit 1
   fi
+}
+
+
+test_iceberg_example_in_pod_using_abfss(){
+  # Test Iceberg integration in Charmed Spark Rock with Azure Storage
+
+  # First create S3 bucket named 'spark'
```
Comment (deusebio) [inline tests/integration/integration-tests.sh]:
```suggestion
  # First create an object bucket named 'spark'
```

File: tests/integration/integration-tests.sh [hunk 4]
```diff
@@ -314,11 +262,61 @@ test_iceberg_example_in_pod(){
...
+  create_azure_container $AZURE_CONTAINER
+
+  # Now, setup S3 properties in service account inside the pod
```
Comment (deusebio) [inline tests/integration/integration-tests.sh]:
```suggestion
  # Now, setup object storage properties in service account inside the pod
```

**Comment by theoctober19th** [comment]
> Great! The tests look really great! thanks!
> 
> **edit** I noticed that sql tests is with S3 only. Would it be possible to make it both for S3 and azure?

Thanks @deusebio. I've just updated the PR to make the SQL tests for both S3 and Azure storage.

**Comment by deusebio** [comment]
great!!! Thanks! Feel free to merge! I'm very happy with this PR, I believe it provides also an improved structure of functionalites!

**Pull request merged**